tt_metal/impl/program/program.cpp

```diff
@@ -835,40 +832,63 @@ void Program::update_runtime_args_transfer_info(Device* device) {
-            this->program_transfer_info.unicast_runtime_args[dst].push_back(transfer_info);
+    if (this->runtime_args_dirty) {
+        this->runtime_args_dirty = false;
+
+        // Unique Runtime Args (Unicast)
+        for (size_t kernel_id = 0; kernel_id < this->num_kernels(); kernel_id++) {
+            auto kernel = detail::GetKernel(*this, kernel_id);
+
+            uint32_t dst = processor_to_l1_arg_base_addr.at(kernel->processor());
+
+            for (const auto& core_coord : kernel->cores_with_runtime_args()) {
+                // can make a vector of unicast encodings here
+                CoreCoord physical_core =
+                    device->physical_core_from_logical_core(core_coord, kernel->get_kernel_core_type());
+                const auto& runtime_args_data = kernel->runtime_args(core_coord);
+                vector<pair<uint32_t, uint32_t>> dst_noc_unicast_info = extract_dst_noc_unicast_info(
+                    detail::GetCoreRangeSet(core_coord).ranges(), kernel->get_kernel_core_type());
+
+                transfer_info_2 transfer_info = {
+                    .dst_base_addr = dst, .dst_noc_info = dst_noc_unicast_info, .linked = false, .data = runtime_args_data};
+                this->program_transfer_info.unicast_runtime_args[dst].push_back(transfer_info);
+            }
+
+            // Common Runtime Args (Multicast)
+            const auto &common_rt_args = kernel->common_runtime_args();
+
+            if (common_rt_args.size() > 0) {
+                uint32_t common_args_addr = dst + kernel->get_common_runtime_args_offset();
+                vector<pair<uint32_t, uint32_t>> dst_noc_multicast_info =
+                    extract_dst_noc_multicast_info<std::vector<CoreRange>>(device, kernel->logical_coreranges(),kernel->get_kernel_core_type());
+                transfer_info_2 transfer_info = {
+                    .dst_base_addr = common_args_addr,
+                    .dst_noc_info = dst_noc_multicast_info,
+                    .linked = false,
+                    .data = common_rt_args};
+                this->program_transfer_info.multicast_runtime_args[common_args_addr].push_back(transfer_info);
+            }
         }
+    } else {
+        for (size_t kernel_id = 0; kernel_id < this->num_kernels(); kernel_id++) {
+            auto kernel = detail::GetKernel(*this, kernel_id);
 
-        // Common Runtime Args (Multicast)
-        const auto &common_rt_args = kernel->common_runtime_args();
+            uint32_t dst = processor_to_l1_arg_base_addr.at(kernel->processor());
 
-        if (common_rt_args.size() > 0) {
-            uint32_t common_args_addr = dst + kernel->get_common_runtime_args_offset();
-            vector<pair<uint32_t, uint32_t>> dst_noc_multicast_info =
-                extract_dst_noc_multicast_info<std::vector<CoreRange>>(device, kernel->logical_coreranges(),kernel->get_kernel_core_type());
```

Comment:
Not applicable to the refactor here, but are Common Runtime Args always same as kernel logical core range?

Reply:
Yes this is true, common runtime args are always applied to kernel->logical_coreranges()